cmd/vm-informant/main.go

```diff
@@ -103,48 +110,106 @@ func main() {
 // We execute ourselves as a subprocess so that it's possible to appropriately cleanup after
 // termination by various signals (or an unhandled panic!). This is worthwhile because we *really*
 // don't want to leave the cgroup frozen while waiting to restart.
-func runRestartOnFailure(args []string, cleanupHooks []func()) {
+func runRestartOnFailure(ctx context.Context, args []string, cleanupHooks []func()) {
 	selfPath := os.Args[0]
-
-	minWaitDuration := time.Second * 5
+	timer := time.NewTimer(0)
+	defer timer.Stop()
 
 	for {
 		startTime := time.Now()
+		processedSignal := make(chan struct{})
+		var exitMode string
 
-		cmd := exec.CommandContext(context.TODO(), selfPath, args...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		err := cmd.Run()
+		func() {
+			defer close(processedSignal)
+			pctx, pcancel := context.WithCancel(context.Background())
+			defer pcancel()
+
+			cmd := exec.Command(selfPath, args...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			err := cmd.Start()
+
+			if err == nil {
+				go func() {
+					select {
+					case <-pctx.Done():
+						return
+					case <-ctx.Done():
+						if cmd.Process == nil || cmd.ProcessState == nil {
+							return
+						}
+						if cmd.ProcessState.ExitCode() >= 0 {
+							return
+						}
```

Comment:
Best I can tell, checking `ProcessState` is not thread-safe. Seems like we're not actually using `ProcessState` other than to check that the command exited — a safer solution could be just setting some atomic boolean when `Wait()` returns.

There's always a possibility of kill failures from TOCTOU issues, but those are unavoidable I think.

Reply:
I think we can use the `pctx` for this purpose.